### PR TITLE
Towards Scala 3 support.

### LINKED ISSUE #16527: Support for Scala3
Now that Scala3 is maturing it would be great to see support for this in the Gradle Scala-plugin. I couldn't find any previously opened issues about it.

### Expected Behavior
Building applications/libraries with Dotty/Scala3 is supported by the Gradle Scala-plugin. Or documented, if this is already supported.

### Current Behavior
There is nothing about Dotty/Scala3 support in the plugin's current documentation.

### Context
I would like to move to scala3 without having to use sbt

### PR DESCRIPTION
**Pull request by dubinsky**
Fixes #16527, #15491

- switch to the latest Zinc;
- detect Scala 3 library dependency;
- configure Scala 3 compiler and bridge as appropriate;
- Scala 3 compiler bridge is distributed as a binary and does not need to be compiled.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by SethTisue** [comment]
are this and #18001 independent efforts?

**Comment by dubinsky** [comment]
> are this and #18001 independent efforts?

Yes, they are.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java [hunk 1]
```diff
@@ -300,6 +300,7 @@ public void execute(ScalaBasePlugin scalaBasePlugin) {
 
                 // exclude the dependencies already provided by SCALA_CONTAINER; prevents problems with Eclipse Scala plugin
                 project.getGradle().projectsEvaluated(gradle -> {
+                    // TODO: support Scala3
```
Comment (tgodzik) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:303]: Eclipse will not currently work anyway :/ so no need to update this.

Comment (dubinsky) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:303]: :)

**Comment by dubinsky** [comment]
The main goal of this  pull request - support for Scala 3 - was not achieved: some classpath manipulations are needed, as demonstrated by #18001, which **does** makes Scala 3 work.

This pull request also attempts to update Zinc to the latest version. This turned out to be:
- more involved than anticipated (one test is still broken);
- not needed to add support for Scala 3 (again, as demonstrated by #18001).
Depending on if, when, and how smoothly #18001 is merged (and released), I might attempt a follow-up pull request for Zinc update (as suggested by @tgodzik in https://github.com/gradle/gradle/pull/18001#issuecomment-897028867) - unless someone who actually knows Zinc beats me to it ;)

Clean-up of the Scala-library-version-recognizing code in the Scala plugin *might* get its own follow-up pull request (again, as suggested by @tgodzik in https://github.com/gradle/gradle/pull/18001#issuecomment-897067536), but this change is less pressing and thus is less likely to happen :)

For the above reasons, this pull request is not relevant. Closing.

**Pull request closed without merging**